e2e/e2eApi.ts

```diff
@@ -0,0 +1,74 @@
+import { Page, Browser, launch } from 'puppeteer'
+import config from '../config'
+
+const serverUrl = `http://${config.server_host}:${config.e2e_port}`
+const launchOptions = {
+  dumpio: false,
+  headless: true,
+  slowMo: 10,
+
+  // Workaround for newPage hang in CircleCI: https://github.com/GoogleChrome/puppeteer/issues/1409#issuecomment-453845568
+  args: [process.env.CI && '--single-process'].filter(Boolean),
+}
+
+export class E2EApiClass {
+  private browser: Browser
+  private page: Page
+
+  public beforeAll = async () => {
+    this.browser = await launch(launchOptions)
+  }
+
+  public beforeEach = async () => {
+    this.page = await this.browser.newPage()
+  }
```

Comment:
As I see we use `browser` only twice: to create/to clsoe pages.

My proposal is to remove `before*` & `after*` logic and pass `page` directly to constructor. Logic of creation `browser` and `page` will be moved to `setup.test.ts`

Reply:
I would rather like to follow Jest semantics with the methods exposed - so that this API will have the same set of methods provided, and it will be straightforward to see where the logic should be put

Reply:
but I see your point - let me try this..

Reply:
yeah - the problem is that logic of initializing `browser` is async - thus we cannot put it to `setup.test.ts` module's initialiser code. And we cannot put it `beforeAll` either, as that will mean that we have to create `global` `e2e` object there - and this is not correct to assign globals in these methods. Of course, we might involve crazy reference-based logic, but I see that it will be far much less obvious and maintainable code.

Thus, lets just have `e2e` object being synchronously created, and make all initialization/uninitialization logic (async one) happening in `beforeAll`/`afterAll` methods - the ones that are designed to handle these aspects.

Reply:
https://jestjs.io/docs/en/setup-teardown#repeating-setup-for-many-tests
Why we can't create a `browser` in async?

> and this is not correct to assign globals in these methods.

Why? Can you point me what is wrong with code below?

```jsx
let browser: Browser
let page: Page

beforeAll(async () => {
  browser = await launch(launchOptions)
})
beforeEach(async () => {
  page = await browser.newPage()
  global['e2e'] = new E2EApiClass(page)
})

afterEach(async () => {
  await page.close()
  global['e2e'] = null
})
afterAll(async () => {
  await browser.close()
})
```

Currently `E2EApiClass` breaks single responsibility and it's obvious because you have `E2EApi` that has another signature. I expect that this class will only provide API, currently it does everything and knows about Jest.

Reply:
done